### PR TITLE
Issue#266 (manual save thumbnail)

### PR DESCRIPTION
**Pull request by rostyslavvnahornyi**

### CODE REVIEW

**Comment by rostyslavvnahornyi** [comment]
@kfarr By the way, I did this task without the firebase plugin, all at once on the frontend code

File: src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx [hunk 1]
```diff
@@ -8,17 +8,73 @@ import { Copy32Icon, Save24Icon } from '../../../icons';
...
   const { currentUser } = useAuthContext();
-  const saveScreenshot = (value) => {
+  const saveScreenshot = async (value) => {
     const screenshotEl = document.getElementById('screenshot');
     screenshotEl.play();
     screenshotEl.setAttribute('screentock', 'type', value);
     screenshotEl.setAttribute('screentock', 'takeScreenshot', true);
+    try {
+      await new Promise((resolve) => setTimeout(resolve, 1000));
+      const aframeScene = document.getElementById('screentock-destination');
```
Comment (kfarr) [inline src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx]: `aframeScene` is actually an img HTML element, can we rename to a more accurate ID for readibility such as `screentockImgElement`

(note to self: screentock-destination is created by toolbar component as img element that receives canvas drawing)

File: src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx [hunk 2]
```diff
@@ -8,17 +8,73 @@ import { Copy32Icon, Save24Icon } from '../../../icons';
...
+      const canvas = await html2canvas(aframeScene);
```
Comment (kfarr) [inline src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx]: - this new `canvas` element is created from an existing image using html2canvas
- I think this is not needed and even if it is, not sure we need a package to do this?
- Instead, see below

File: src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx [hunk 3]
```diff
@@ -8,17 +8,73 @@ import { Copy32Icon, Save24Icon } from '../../../icons';
...
+      const resizedCanvas = document.createElement('canvas');
+      resizedCanvas.width = 320;
+      resizedCanvas.height = 240;
+      const context = resizedCanvas.getContext('2d');
+
+      context.drawImage(canvas, 0, 0, 320, 240);
```
Comment (kfarr) [inline src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx]: Instead of converting img element to canvas via html2canvas and then drawing in another canvas; in theory we can pass the screentock-destination img directly. (`drawImage()` should accept `HTMLImageElement`)

So this could be:
`     context.drawImage(aframeScene, 0, 0, 320, 240);`

File: src/components/scenegraph/Toolbar.js [hunk 4]
```diff
@@ -147,11 +147,15 @@ export default class Toolbar extends Component {
       }
 
       // we want to save, so if we *still* have no sceneID at this point, then create a new one
-      if (!currentSceneId || !!doSaveAs) {
+      if (!currentSceneId || (!!doSaveAs && !isCurrentUserTheSceneAuthor)) {
```
Comment (kfarr) [inline src/components/scenegraph/Toolbar.js]: I don't understand why these changes are related to thumbnail upload (#266)?

File: src/components/scenegraph/Toolbar.js [hunk 5]
```diff
@@ -147,11 +147,15 @@ export default class Toolbar extends Component {
...
         console.log(
           'no urlSceneId or doSaveAs is true, therefore generate new one'
         );
+
         currentSceneId = await generateSceneId(this.props.currentUser.uid);
+        if (!doSaveAs) {
+          window.location.hash = `#/scenes/${currentSceneId}.json`;
+        }
```
Comment (kfarr) [inline src/components/scenegraph/Toolbar.js]: I don't understand why these changes are related to thumbnail upload (#266)?

File: src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx [hunk 6]
```diff
@@ -8,17 +8,73 @@ import { Copy32Icon, Save24Icon } from '../../../icons';
 import { loginHandler } from '../SignInModal';
 import { useAuthContext } from '../../../contexts';
 import Toolbar from '../../scenegraph/Toolbar';
+import { db, storage } from '../../../services/firebase';
+import html2canvas from 'html2canvas';
```
Comment (kfarr) [inline src/components/modals/ScreenshotModal/ScreenshotModal.component.jsx]: can we try to not have this dependency? see below

**Review by kfarr — CHANGES_REQUESTED** [review]

**Comment by kfarr** [comment]
To integrate:
- [x] remove altered Save As... logic from Toolbar.js -- is this related to another ticket / PR?
- [x] allow firing of this event with a button per figma design (may require separating the code such that it can be fired on demand and not with download as...)
- [x] remove firing of this event when just downloading a file
- [x] security rules check: as a user, I should not be able to upload a new scene thumbnail to a scene to which I am not an owner
- [x] current scene for URL path is determined by wrong method (url bar instead of metadata)
- [x] move the new STREET.utils from index.html into json-utils (and create ticket https://github.com/3DStreet/3dstreet/issues/390 to convert other global functions to STREET.utils)
- [x] add success / error notify messages
- [x] investigate save as error
- [ ] merge!


picture1:
image

**Review by kfarr — APPROVED** [review]
